server/src/main/java/org/elasticsearch/transport/InboundDecoder.java

```diff
@@ -216,10 +215,14 @@ static void checkHandshakeVersionCompatibility(TransportVersion handshakeVersion
         }
     }
 
-    static void checkVersionCompatibility(TransportVersion remoteVersion, TransportVersion currentVersion) {
-        if (remoteVersion.isCompatible(currentVersion) == false) {
+    static void checkVersionCompatibility(TransportVersion remoteVersion) {
+        if (TransportVersion.isCompatible(remoteVersion) == false) {
```

Comment:
woudl `remoteVersion.isCompatible()` read better?

Reply:
Possibly, it's comparing against the static `MINIMUM_COMPATIBLE` constant, given that TransportVersion instances represents specific versions, it'll be a bit weird to have an instance method that actually refers to a static constant.

Keeping it as a static method reinforces the fact that it's comparing against the static running version of ES, which is held in static constants.

Reply:
it looks odd to use a static utility method instead of instead instance method, when there is no need for this.
Now it is just an utility wrapper around `version.onOrAfter(MINIMUM_COMPATIBLE)`  which is an instance method. 

To me the use of static inside an an instance method is fine, it is an implementation detail.

Reply:
To me its all about what the method is actually acting on. An instance call `version.isCompatible()` is asking the instance to determine 'are you compatible'? But its not really to be determined by the version instance, its to be determined _by the system_, as a static property of the running code. So its up to the _system_ to determine compatibility, not the _version instance_.

Hence a static method, asking the system 'is this version instance compatible'?

Reply:
> An instance call version.isCompatible() is asking the instance to determine 'are you compatible'? But its not really to be determined by the version instance, its to be determined by the system

`TransportVersion.isCompatible(remoteVersion)` does the same thing as `remoteVersion.isCompatible `. The use of static constant is an implementation detail which might change if we want to. 
If a `system` was required to determine compatibility I would expect the signature look like `remoteVersion.isCompatible(system)`. But I don't really see the `system` here.. 

`TransportVersion.isCompatible(remoteVersion)` is asking for encapsulation to be broken. What if compatibility requires some private parts of `remoteVersion` to be exposed? 

I like the first part where you mentioned `version.isCompatible() is asking the instance to determine 'are you compatible'` This read good to me.

see similar examples of the same pattern to be modeled
Version.isLegacyIndexVersion
or RestApiVersion.minimumSupported (and similar other methods)

Reply:
The previous implementation of `isCompatible` was reflexive, this one is not, as it is comparing with static variables set at compile time, not comparing two instances. My point is that it is _not up to the instance_ to determine whether it is compatible, as the same logical instance on two different systems can have different results. It's up to the system (here, static constants in `TransportVersion` set at compile time), to determine whether a particular instance is compatible with the code running on the system.

Reply:
The context of the question is the running system, with  its `MINIMUM_COMPATIBLE` constant, I think it makes sense that the running code asks if the `remoteVersion` is compatible.  The implementation of `isCompatible` is odd, I would have expected it to be a static call, eg `MINIMUM_COMPATIBLE.onOrBefore(version)`, rather than supplying a static parameter.